res/controllers/Numark-Scratch-scripts.js

```diff
@@ -0,0 +1,452 @@
+        break;
+    case 0x7F:  // Picnic Bench/Fast Cut
+        engine.setValue("[Mixer Profile]", "xFaderMode", 0);
+        engine.setValue("[Mixer Profile]", "xFaderCalibration", 0.9);
+        engine.setValue("[Mixer Profile]", "xFaderCurve", 7.0);
+    }
+};
+
+NumarkScratch.crossfader.xFaderReverse = function(channel, control, value, _status, _group) {
+    // 0x7F is ON, 0x00 is OFF
+    engine.setValue("[Mixer Profile]", "xFaderReverse", (value === 0x7F) ? 1 : 0);
+};
+
+NumarkScratch.setChannelInput = function(channel, control, value, _status, _group) {
+    var number = (control === 0x57) ? 1 : 2;
+    var channelgroup = "[Channel" + number + "]";
+
+    switch (value) {
+    case 0x00:  // PC and turn on vinyl control
+        engine.setValue(channelgroup, "passthrough", 0);
+        engine.setValue(channelgroup, "vinylcontrol_enabled", 1);
+        break;
+    case 0x02:  // PHONO/LINE and turn off vinyl control
+        engine.setValue(channelgroup, "passthrough", 1);
+        engine.setValue(channelgroup, "vinylcontrol_enabled", 0);
+        break;
+    }
+};
+
+NumarkScratch.Deck = function(number) {
+    components.Deck.call(this, number);
+
+    var channel = number - 1;
+    //var deck = this;
+
+    this.pflButton = new components.Button({
+        midi: [0x90 + channel, 0x1B],
+        key: "pfl",
+        output: function(value) {
+            var note = (value === 0x00 ? 0x80 : 0x90) + channel;
+            midi.sendShortMsg(note, 0x1B, this.outValueScale(value));
+        }
+    });
+
+    this.volume = new components.Pot({
+        inKey: "volume"
+    });
+
+    this.treble = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter3"
+    });
+
+    this.mid = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter2"
+    });
+
+    this.bass = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter1"
+    });
+
+    this.filter = new components.Pot({
+        group: "[QuickEffectRack1_" + this.currentDeck + "]",
+        inKey: "super1"
+    });
+
+    this.gain = new components.Pot({
+        inKey: "pregain"
+    });
+
+    this.loopEncoder = new components.Encoder({
+        unshift: function() {
+            if (!NumarkScratch.invertLoopEncoderFunction) { //if set to false then normal function
+                this.input = function(channel, control, _value, _status, _group) {
+                    var loopSize = engine.getValue(this.group, "beatloop_size");
+                    if (control === 52 && loopSize >= (1/16)) { //turn left
+                        engine.setValue(this.group, "beatloop_size", loopSize / 2);
+                    } else if (control === 53 && loopSize < 256) { // turn right
+                        engine.setValue(this.group, "beatloop_size", loopSize * 2);
+                    }
+                };
```

Comment:
you can greatly simplify this by creating different components for the different controls and then using the `loop_halve`/`loop_double` COs.

Reply:
@Swiftb0y this is one of the issues I'm having. Abstracting this into a different component and then using COs is proving challenging.
This long way round was how I could make logical make sense of what I wanted as result
Do you have an example I could look at for inspiration?

Reply:
This is essentially what I was referring to. I also created the `possiblyInvert` function to conditionally assign the correct function (you can see that I saved a lot of duplication there). I recommend you make the `possiblyInvert` function a global utility function and then use the same technique to reduce the duplication for the other components that use `NumarkScratch.invertLoopEncoderFunction` (or just remove that feature entirely). 
```javascript
var possiblyInvert = function(key) {
    if (NumarkScratch.invertLoopEncoderFunction) {
        return key === "unshift" ? "shift" : "unshift";
    }
    return key;
}
var loopPlusButtonConf = {
    midi: [0x90, 0x34],
};
loopPlusButtonConf[possiblyInvert("unshift")] = function() {
    this.group = "[Channel" + channel + "]"
    this.inKey = "loop_halve";
};
loopPlusButtonConf[possiblyInvert("shift")] = function() {
    this.group = "[Library]";
    this.inKey = "MoveDown";
};

this.loopPlus = new components.Button(loopPlusButtonConf);

// TODO same as above just with loopMinus now. Then map the loopPlus "Button" for control 0x35 (53) and loopMinus for 0x34 (52)
```